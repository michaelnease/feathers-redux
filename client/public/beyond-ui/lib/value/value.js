const UNDEFINED = (void 0);

/*
 * `isUndefined` checks if a value is `undefined`. It checks inputs strictly
 * and returns `false` for any other value.
 */
const isUndefined = v => v === UNDEFINED;

/*
 * `isString` checks if a value is string.
 */
const isString = v => typeof v === "string";

const isBoolean = v => typeof v === "boolean";

/*
 * `isObject` checks if a value is a plain object
 */
const isObject = v => Object.prototype.toString.call(v) === "[object Object]";

/*
* `isSet` checks if a value is a Set.
*/
const isSet = v => Object.getPrototypeOf(v) === Set.prototype;

/*
 * `isFunction` checks if a value is a function
 */
const isFunction = v => typeof v === "function";

/*
 * `isNumber` checks if a value is a number. Useful for when
 * you want to check if something is a number, and you don't
 * care about NaN. If you want to check for NaN need to use
 * the standard Number.isNaN.
 */
const isNumber = v => typeof v === "number" && !Number.isNaN(v);

/*
* `isArray` checks if a value is Array.
*/
const isArray = v => Array.isArray(v);

/*
* `intersect` checks if two sets intersect and returns
* the intersect set. If the inputs are arrays, they will
* be turned into sets. The function only accepts either
* both arrays, or both sets.
*/
const intersect = (a, b) => {
  if (![a, b].every(isArray)) {
    if (![a, b].every(isSet)) {
      throw new Error("Both arguments should be either both Sets or Arrays");
    }
  }

  let set1;
  let set2;

  if ([a, b].some(isArray)) {
    set1 = new Set(a);
    set2 = new Set(b);
  } else {
    set1 = a;
    set2 = b;
  }

  return new Set([...set1].filter(x => set2.has(x)));
};

/*
 * Common error types, useful for keeping the error values consistent.
 */
const ERRORS = {
  BAD_ARGUMENT: "BAD_ARGUMENT"
};

/* a noop function, useful for setting default values */
const noop = () => UNDEFINED;

/*
 * Error returns an object with a details field, useful for
 * indicating that there was an error. The `type` and the `message` should
 * both be provided as strings.
 * For more detailed usage cases, please take a look at the test cases.
 * Prefer using the constructor versions instead, but this is useful
 * if you want to return a value and have more control over error behavior.
 */
const error = (type, details) => {
  const isValidParam = [type, details].every(isString);
  if (!isValidParam) {
    throw new Error("Type and details should be both strings.");
  }

  /* Define to primitive and coercion behaviors */
  return {
    details,
    type,
    [Symbol.toPrimitive](hint) {
      if (hint === "string") {
        return type;
      }

      if (hint === "number") {
        return -1;
      }

      return type;
    }
  };
};

/* Custom error types, extending built-in Error */
class BadArgument extends Error {
  constructor(message) {
    super(message);
    this.name = ERRORS.BAD_ARGUMENT;
    this.message = message;
  }

  toJSON() {
    return {
      error: {
        name: this.name,
        message: this.message,
        stacktrace: this.stack
      }
    };
  }
}

/*
 * Makes a plain object immutable. Useful for preventing
 * others from changing properties and values by accident.
 */
const deepFreeze = object => {
  const propNames = Object.getOwnPropertyNames(object);
  for (const name of propNames) {
    const value = object[name];

    object[name] =
      value && typeof value === "object" ? deepFreeze(value) : value;
  }

  return Object.freeze(object);
};

module.exports = {
  UNDEFINED,
  isUndefined,
  isString,
  isBoolean,
  isObject,
  isFunction,
  intersect,
  isArray,
  isSet,
  isNumber,
  error,
  ERRORS,
  BadArgument,
  deepFreeze,
  noop
};
