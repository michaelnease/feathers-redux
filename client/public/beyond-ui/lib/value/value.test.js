import * as value from "./value";

describe("`isUndefined` should:", () => {
  it("return true if input is not given.`", () => {
    expect(value.isUndefined()).toBe(true);
  });

  it("return true if input is value `undefined`.", () => {
    expect(value.isUndefined(void 0)).toBe(true);
  });

  it("return true if input is of type undefined.", () => {
    expect(value.isUndefined(value.UNDEFINED)).toBe(true);
  });

  it("return false if input is `null`.", () => {
    expect(value.isUndefined(null)).toBe(false);
  });
});

describe("`isString` should:", () => {
  it("return true if input arg is a string", () => {
    expect(value.isString("boo")).toBe(true);
  });

  it("return true if input arg is a string.", () => {
    expect(value.isString("")).toBe(true);
  });

  it("return false if input arg is not a string", () => {
    expect(value.isString({})).toBe(false);
  });
});

describe("`error` should:", () => {
  it("throw an error if no input is given", () => {
    expect(() => {
      value.error();
    }).toThrow();
  });

  it("return an object with details and type for given params.", () => {
    const e = value.error("INVALID_AUTH", "Invalid auth token.");
    expect(typeof e.details).toBe("string");
    expect(typeof e.type).toBe("string");
  });

  it("throw an error if both type and details are not given.", () => {
    expect(() => {
      value.error("AUTH_ERROR");
    }).toThrow();
  });

  it("should return the type value in string context", () => {
    const e = value.error("AUTH_ERROR", "Invalid auth token");
    expect(/AUTH_ERROR/.test(e)).toBe(true);
  });

  it("should return the type value in default context", () => {
    const e = value.error("AUTH_ERROR", "Invalid auth token");
    expect(e + 5).toBe("AUTH_ERROR5");
  });

  it("should return -1 in number context", () => {
    const e = value.error("AUTH_ERROR", "Invalid auth token");
    expect(e - 1).toBe(-2);
  });
});

describe("`isNumber` should:", () => {
  it("return true if input param is a number", () => {
    expect(value.isNumber(25)).toBe(true);
  });

  it("return false if input param is NaN", () => {
    expect(value.isNumber(NaN)).toBe(false);
  });

  it("return false if input param is not a number", () => {
    expect(value.isNumber("Hello")).toBe(false);
  });
});

describe("`isObject` should:", () => {
  it("return true if input param is an object", () => {
    expect(value.isObject({})).toBe(true);
  });

  it("return false if input param is not object", () => {
    expect(value.isObject("boo")).toBe(false);
  });

  it("return false if input param is not object", () => {
    expect(value.isObject(null)).toBe(false);
  });

  it("return false if input param is not object", () => {
    function Hello() {}
    expect(value.isObject(Hello)).toBe(false);
  });

  it("return false if input param is not object", () => {
    expect(value.isObject([])).toBe(false);
  });
});

describe("`isFunction` should:", () => {
  it("return true if input param is a function", () => {
    expect(value.isFunction(() => {})).toBe(true);
  });

  it("return false if input param not a function", () => {
    expect(value.isFunction("hello")).toBe(false);
  });
});

describe("`errorTypes` should:", () => {
  it("have badArg defined", () => {
    expect(value.ERRORS.BAD_ARGUMENT).toBeDefined();
  });
});

describe("`BadArg` should:", () => {
  it("construct a custom BadArg error type with a custom error name", () => {
    expect(new value.BadArgument().name).toBe(value.ERRORS.BAD_ARGUMENT);
  });

  it("construct a custom badarg error with an optional message if given", () => {
    expect(new value.BadArgument("boo").message).toBe("boo");
  });

  // it("serialize correctly when stringified", () => {
  //   const err = new value.BadArgument("boo");
  //   const serialized = JSON.stringify(err);
  //   expect(serialized).toMatchSnapshot();
  // });
});

it("isArray should return true if input is array", () => {
  expect(value.isArray([])).toBe(true);
});

it("isSet should return true if input is a Set", () => {
  expect(value.isSet(new Set())).toBe(true);
});

describe("intersection should return an intersection set", () => {
  it("should throw error if both args are not arrays", () => {
    expect(() => {
      value.intersect([1], "x");
    }).toThrow();
  });

  it("should throw error if both args are not sets", () => {
    expect(() => {
      value.intersect(new Set(), "x");
    }).toThrow();
  });

  it("should return intersect two arrays", () => {
    const intersectionSet = value.intersect([1, 2], [2, 3]);
    expect([...intersectionSet]).toEqual([2]);
  });

  it("should intersect two sets", () => {
    const s1 = new Set([1, 2]);
    const s2 = new Set([2, 4]);
    const output = value.intersect(s1, s2);
    expect([...output]).toEqual([2]);
  });
});
