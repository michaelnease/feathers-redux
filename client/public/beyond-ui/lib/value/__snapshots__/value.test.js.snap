// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`BadArg\` should: serialize correctly when stringified 1`] = `"{\\"error\\":{\\"name\\":\\"BAD_ARGUMENT\\",\\"message\\":\\"boo\\",\\"stacktrace\\":\\"BAD_ARGUMENT: boo\\\\n    at Object.<anonymous> (/Users/ameyghani/projects/beyond/roa/roa-ui/apiserver/src/shared/value/value.test.js:133:17)\\\\n    at Object.asyncJestTest (/Users/ameyghani/projects/beyond/roa/roa-ui/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)\\\\n    at resolve (/Users/ameyghani/projects/beyond/roa/roa-ui/node_modules/jest-jasmine2/build/queueRunner.js:43:12)\\\\n    at new Promise (<anonymous>)\\\\n    at mapper (/Users/ameyghani/projects/beyond/roa/roa-ui/node_modules/jest-jasmine2/build/queueRunner.js:26:19)\\\\n    at promise.then (/Users/ameyghani/projects/beyond/roa/roa-ui/node_modules/jest-jasmine2/build/queueRunner.js:73:41)\\\\n    at process._tickCallback (internal/process/next_tick.js:68:7)\\"}}"`;
